PromQL: Add `fill*()` binop modifiers to provide default values for missing series

juliusv opened the pull request:
This starts implementing the [proposal for default value binop modifiers](https://github.com/prometheus/proposals/blob/main/proposals/0070-binop-fill-modifier.md) throughout the stack, from the PromQL parser+engine all the way to the UI and PromLens visualizer.

So you can now do something like:

```
# Fill in missing series from either side with a default value of 0.
metric1 + fill(0) metric2
```

Or:

```
# Fill in missing custom rate thresholds with a default value of 23.
some_rates > fill_right(23) some_rate_thresholds
```

A few special considerations to note:

* Other binop modifiers like `on()`, `group_left()`, etc. are reserved words that are NOT allowed as metric names. I did not want to break existing users, so I could not adopt the same behavior for the new fill modifiers. For example, a user might have an expression like `foo + fill`, where both are metric names, and that still needs to be parseable as a metric name. Thus I had to extend the lexer to only emit fill modifier tokens if there is a parenthesis following them. Maybe there's a better way of doing this directly in the grammar, but I think yacc works makes that hard? Input / experimentation by others welcome.
* I am not implementing the [parameterless variants](https://github.com/prometheus/proposals/blob/main/proposals/0070-binop-fill-modifier.md#possible-parameter-less-variants-of-the-modifiers) for now, so you always have to include the parentheses with a value inside.
* The current binOp fill implementation in the engine introduces a closure function in a place that might be performance-sensitive. It was the easiest way to implement things for now, but we need to benchmark this properly and determine if this is ok or whether the binop evaluation should be changed in such a way that it is centered around the match label hashbucket (similar to how Thanos does it), with both sides (LHS + RHS) being filled in for each bucket before performing the final binop operation (or filling in of missing series).

Other than those notes, progress is roughly:

* [X] Add the `fill`, `fill_left`, and `fill_right` modifiers to the PromQL lexer + parser.
* [X] Implement the evaluation logic for the new modifiers in the PromQL engine.
* [X] Update the Go-based PromQL printer / formatter to correctly render the new modifiers.
* [X] Update the PromQL grammar for the CodeMirror code editor to recognize the new modifiers.
* [X] Add autocompletion support for the new modifiers in the PromQL editor.
* [ ] Add linting support for the new modifiers in the PromQL editor. (HALF DONE, can be completed later)
* [X] Support the new modifier in the PromLens-style tree view.
* [X] Support visualizing the new modifier behavior in the "Explain" tab for binary operator nodes.
* [X] Add unit tests and integration tests to cover the new functionality.
* [X] Update the Prometheus documentation to include the new modifiers and their usage.

Here's a screenshot of how it looks in the PromLens tree view and the matching visualization in the "Explain" tab:

<img width="3813" height="1864" alt="image" src="https://github.com/user-attachments/assets/42937498-c324-46ba-aad7-52e680356495" />

Fixes https://github.com/prometheus/prometheus/issues/13625

#### Does this PR introduce a user-facing change?

```release-notes
[FEATURE] PromQL: Add `fill()` / `fill_left()` / `fill_right()` binop modifiers for specifying default values for missing series
```

juliusv committed:
PromQL: Add `fill*()` binop modifiers to provide default values for missing series

Signed-off-by: Julius Volz <julius.volz@gmail.com>

vpranckaitis commented on promql/engine.go:
```diff
@@ -2929,7 +2922,7 @@ func (ev *evaluator) VectorBinop(op parser.ItemType, lhs, rhs Vector, matching *
 			}
 		case !keep:
-			continue
+			return
```
Also, judging by the fact that comparison without `bool` modifier returns the metric name of LHS, binary comparison operation seems to be treated as a way to filter LHS timeseries. However, the test above shows appearance of `{case="2"}` and `{case="4"}` timeseries, which are coming from RHS. Not sure if this is the intended behavior. 

This is what causes the second test case to fail. Because `bool` modifier requires metric names to be dropped, you end up with `metric_a{case="2"}` and `{case="2"}` colliding once delayed name removal is applied.

vpranckaitis replied:
I had this idea, and I want to share it before it becomes way more expensive to make changes. Just throwing it out, do what you want with it.

It's related to this quote:

> Now semantically, I don't think filling in the LHS in a filtering operation makes much sense most of the time. Usually you'll only want to fill in the RHS with some default value using `fill_right`.

People have discussed a feature where `fill` without arguments would automatically select sensible default values. What if similarly, `fill(x)` would not just be a shorthand of `fill_left(x) fill_right(x)`, but it would select sensible sides where to put default values?  And you would still have `fill_left()` and `fill_right()` for ultimate control.

If we can quite confidently say that some `fill()` use cases don't make much sense and a user should know better to do something else, then maybe the PromQL could do the right thing by default? And if you don't like the default, you can still do whatever you like with `fill_left()` and `fill_right()`.

I can think of a few cases where defaults on both sides might not make much sense:
1. Already discussed comparison operation between two vectors, which works like a filter for LHS;
2. `group_left`/`group_right`, which takes everything from the target side and adjust the values and labels using the data on the opposite side.

For both of these, there are cases where you end up with some timeseries having fewer labels, because some labels come to the results just from one of the sides and you might need to insert a default value to that side. This shouldn't be a problem with `fill()` if it fills just one, the sensible side (though it would still happen with `fill_left/right()`).

From UX perspective, this might also make it feel that we're not just inserting/filling some default values, but we're actually integrating into existing PromQL matching, and extending how it works (as hinted in [this message](https://github.com/prometheus/prometheus/issues/13625#issuecomment-1984049276) from the related issue).

But I want to reiterate, I'm just throwing this idea out and feel free to do what you want with it.

juliusv committed:
Add fill modifier PromQL tests

Signed-off-by: Julius Volz <julius.volz@gmail.com>

juliusv committed:
Add PromLens binop matching explain view tests

Signed-off-by: Julius Volz <julius.volz@gmail.com>

juliusv committed:
Document new fill binop modifiers

Signed-off-by: Julius Volz <julius.volz@gmail.com>

juliusv committed:
Add new fill modifiers to features test data

Signed-off-by: Julius Volz <julius.volz@gmail.com>

juliusv commented:
I added tests and documentation for the modifiers now and removed the `[WIP]` from the PR description, so I think it's ready for broader review.

For the closure that I introduced in the binop computation, it seems fine, or at least it doesn't seem to change the `BenchmarkJoinQuery` benchmark at all in terms of allocs or speed.

Before:

```
BenchmarkJoinQuery/expr=rpc_request_success_total_+_rpc_request_error_total/steps=10000-20         	       1	2612443093 ns/op	1580342576 B/op	  577780 allocs/op
BenchmarkJoinQuery/expr=rpc_request_success_total_+_ON_(job,_instance)_GROUP_LEFT_rpc_request_error_total/steps=10000-20         	       1	3083784580 ns/op	1977043056 B/op	20567736 allocs/op
BenchmarkJoinQuery/expr=rpc_request_success_total_AND_rpc_request_error_total{instance=~"0.*"}/steps=10000-20                    	       2	 521685498 ns/op	39702924 B/op	  306865 allocs/op
BenchmarkJoinQuery/expr=rpc_request_success_total_OR_rpc_request_error_total{instance=~"0.*"}/steps=10000-20                     	       1	1077586308 ns/op	518800192 B/op	  311608 allocs/op
BenchmarkJoinQuery/expr=rpc_request_success_total_UNLESS_rpc_request_error_total{instance=~"0.*"}/steps=10000-20                 	       2	 986880906 ns/op	39771716 B/op	  306864 allocs/op
```

After:

```
BenchmarkJoinQuery/expr=rpc_request_success_total_+_rpc_request_error_total/steps=10000-20         	       1	2619260793 ns/op	1580348944 B/op	  577797 allocs/op
BenchmarkJoinQuery/expr=rpc_request_success_total_+_ON_(job,_instance)_GROUP_LEFT_rpc_request_error_total/steps=10000-20         	       1	3126463664 ns/op	1977042880 B/op	20567734 allocs/op
BenchmarkJoinQuery/expr=rpc_request_success_total_AND_rpc_request_error_total{instance=~"0.*"}/steps=10000-20                    	       2	 509301824 ns/op	39703108 B/op	  306866 allocs/op
BenchmarkJoinQuery/expr=rpc_request_success_total_OR_rpc_request_error_total{instance=~"0.*"}/steps=10000-20                     	       1	1054013104 ns/op	518800112 B/op	  311607 allocs/op
BenchmarkJoinQuery/expr=rpc_request_success_total_UNLESS_rpc_request_error_total{instance=~"0.*"}/steps=10000-20                 	       2	 971900134 ns/op	39771844 B/op	  306863 allocs/op
```

A remaining issue is the interplay of delayed metric name removal (`--enable-feature=promql-delayed-name-removal`) and fill modifiers that @vpranckaitis pointed out: If a series exists at some time steps and not others, it will exist with its metric name (`metric_a{case="2"}`) when it is present and without the metric name (`{case="2"}`) when it has to be filled in. This is no problem at each single resolution step (as only one variant of the series will be present at any given time, either the real one or the filled-in one), but apparently the output vectors from multiple resolution steps are gathered into a single result matrix, and with delayed metric name removal, the metric name is removed only after that matrix has been gathered, causing collisions at that point in the process, when `metric_a{case="2"}` gets reduced to `{case="2"}`, which already exists in the matrix. Any advice on how to correctly address this is welcome.

bboreham commented:
I believe Jorge is on paternity leave for a few months, so may not reply.

juliusv commented:
@roidelapluie Yes, I've noticed now that the problem is actually more general and does not require the new fill modifiers to be provoked. The problem is in general that we:

* First gather all resolution steps over time in a big matrix,
* ...only then apply modifications to that matrix (removing the metric name) that could result in some series names colliding across steps
* ...and only then detecting that as a collision in `ContainsSameLabelset`: https://github.com/prometheus/prometheus/blob/583bc01cc9f37bed237880083d44cab0d01e056a/promql/value.go#L321-L341

So you can also provoke it like this:

```
load 10m
    metric_a   1  _
    metric_b   3  4

eval range from 0 to 20m step 10m -metric_a or -metric_b
    {} -1  -4
```

Or if you want to provoke it with the new fill modifiers, you could do:

```
load 10m
    metric_a   1  _
    metric_b   _  4

eval range from 0 to 20m step 10m metric_a <= bool fill(0) metric_b
    {} 0  1
```

There's also more ways to provoke this, like:

```
load 10m
    metric_a   1  _
    metric_b   _  4

eval range from 0 to 20m step 10m -metric_a or on(__name__) -metric_b
    {} -1  -4
```

All these queries run fine without delayed metric name removal, but cause collision errors when delayed metric name removal is enabled. I mentioned the same example queries in https://github.com/prometheus/prometheus/issues/15855 already as well. So I think this is generally an issue that can and should be fixed separately from the new fill modifiers. I guess you'd either have to:

* In the big final matrix, merge together series with the same label set that don't have overlapping sample values at the same timestamp.
* Or do any transformations that could cause collisions before gathering everything into one big final matrix.

roidelapluie commented:
WDYT of https://github.com/prometheus/prometheus/pull/17678 ?

linasm commented:
Did a basic search, did not find anything, perhaps I has missed something - my question is:
was there a discussion on whether this should be rolled out as an experimental feature first, or is is going straight for GA?

juliusv commented:
@linasm I wanted to bring up that question today as well. While I am tempted to just merge it without a flag immediately, I think putting it behind an feature flag might actually be the safer and more responsible way to go. So far it feels like only a small handful of people have taken a look at the details, and it's probably best to not set things 100% in stone yet. You probably would agree?

juliusv committed:
Put binop fill modifiers behind a feature flag

Signed-off-by: Julius Volz <julius.volz@gmail.com>

bwplotka reviewed (approved):
Looks good to me.

The only nit is around test case for the metric names like `"fill + fill" that hits the special case we added.

juliusv committed:
Add a few fill modifier tests with keyword-like metric names

Signed-off-by: Julius Volz <julius.volz@gmail.com>

juliusv committed:
Fix a missing space/newline in the binop docs

Signed-off-by: Julius Volz <julius.volz@gmail.com>

juliusv commented on docs/querying/operators.md:
```diff
@@ -311,6 +308,58 @@ left:
+are not included in the result vector. Fill modifiers allow overriding this behavior by filling
+in missing series on either side of a binary operation with a provided default sample value:
+
+* `fill(<value>)`: Fill in missing matches on either side with `value`.
```
Hehe yeah, I thought of that and a few other options, but arrived at the same conclusion that it would be too adventurous and inconsistent with the modifiers we already have.

colega commented:
@juliusv you might want to update the status in https://github.com/prometheus/proposals/blob/main/proposals/0070-binop-fill-modifier.md

juliusv commented:
@colega Good point, created https://github.com/prometheus/proposals/pull/88
